lib/Sabberworm/CSS/Parser.php

```diff
@@ -64,20 +71,91 @@ public function __construct($sText, Settings $oParserSettings = null, $iLineNo =
+			$this->sOriginalCharset = 'utf-32be';
+		} else if (strpos($this->sText, self::BOM32LE) === 0) {
+			$this->sText = substr($this->sText, strlen(self::BOM32LE));
+			$this->sOriginalCharset = 'utf-32le';
+		} else if (strpos($this->sText, self::BOM16BE) === 0) {
+			$this->sText = substr($this->sText, strlen(self::BOM16BE));
+			$this->sOriginalCharset = 'utf-16be';
+		} else if (strpos($this->sText, self::BOM16LE) === 0) {
+			$this->sText = substr($this->sText, strlen(self::BOM16LE));
+			$this->sOriginalCharset = 'utf-16le';
+		} else if (preg_match('/(.*)@charset\s+["\']([a-z0-9-]+)["\']\s*;/ims', $this->sText, $aMatches)) {
+			// This is a simplified guessing, the charset atRule location is validated later,
+			// hopefully this is not used much these days.
+			if (trim($aMatches[1]) === '' and preg_match('/^@charset\s+["\']([a-z0-9-]+)["\']\s*;/im', $aMatches[0])) {
+				$this->sOriginalCharset = strtolower($aMatches[2]);
+			}
+		}
+
+		// Convert all text to utf-8 so that code does not have to deal with encoding conversions and incompatible characters.
+		if ($this->sOriginalCharset !== 'utf-8') {
+			if (function_exists('mb_convert_encoding')) {
+				$this->sText = mb_convert_encoding($this->sText, 'utf-8', $this->sOriginalCharset);
+			} else {
+				$this->sText = iconv($this->sOriginalCharset, 'utf-8', $this->sText);
+			}
+		}
 
-	public function setCharset($sCharset) {
-		$this->sCharset = $sCharset;
-		$this->aText = $this->strsplit($this->sText);
-		$this->iLength = count($this->aText);
+		// Multibyte support can make the parsing slower,
+		// but even if it is disabled the unicode characters usually survive this parsing unharmed.
+		$this->sTextLibrary = 'ascii';
+		if (!$this->oParserSettings->bMultibyteSupport) {
+			$this->iLength = strlen($this->sText);
+			return;
+		}
+
+		// If there are only ASCII characters in the CSS then we can safely use good old PHP string functions here.
+		if (function_exists('mb_convert_encoding')) {
+			$sSubst = mb_substitute_character();
+			mb_substitute_character('none');
+			$asciiText = mb_convert_encoding($this->sText, 'ASCII', 'utf-8');
+			mb_substitute_character($sSubst);
+		} else {
+			$asciiText = @iconv('utf-8', 'ASCII//IGNORE', $this->sText);
+		}
+		if ($this->sText !== $asciiText) {
+			if (function_exists('mb_convert_encoding')) {
+				// Usually mbstring extension is much faster than iconv.
+				$this->sTextLibrary = 'mb';
+			} else {
+				$this->sTextLibrary = 'iconv';
+			}
+		}
+		unset($asciiText);
+		$this->iLength = $this->strlen($this->sText);
+
+		// Substring operations are slower with unicode, aText array is used for faster emulation.
+		if ($this->sTextLibrary !== 'ascii') {
+			$this->aText = preg_split('//u', $this->sText, null, PREG_SPLIT_NO_EMPTY);
+			if (!is_array($this->aText) || count($this->aText) !== $this->iLength) {
+				$this->aText = null;
+			}
+		}
 	}
 
 	public function getCharset() {
-		return $this->sCharset;
+		return 'utf-8';
```

Comment:
I kept the getChartset() for backwards compatibility because it was telling you in which encoding are the data structures - now it is always utf-8, so all code should work fine as before, I think it is better to keep it this way in case somebody relies on it, alternatively it could be removed completely or at least deprecated